src/Toolchain.ml

```diff
@@ -0,0 +1,647 @@
+                   Js.log3 "Invalid JSON file found. Ignoring..." manifest
+                     manifestFile;
+                   None |> Js.Promise.resolve)
+        | ".opam" ->
+          Fs.readFile manifestFile
+          |> Js.Promise.then_ (function
+               | "" -> None |> Js.Promise.resolve
+               | _ -> Some (Opam, projectRoot) |> Js.Promise.resolve)
+        | _ -> None |> Js.Promise.resolve )
+
+    let lookup projectRoot =
+      Fs.readDir (Fpath.toString projectRoot)
+      |> Js.Promise.then_ (function
+           | Error msg -> Js.Promise.resolve (Error msg)
+           | Ok files ->
+             files
+             |> Js.Array.map (parse projectRoot)
+             |> Js.Promise.all
+             |> Js.Promise.then_ (fun l ->
+                    Ok
+                      ( Js.Array.reduce
+                          (fun acc x ->
+                            Js.Array.concat acc
+                              ( match x with
+                              | Some x -> Array.of_list [ x ]
+                              | None -> [||] ))
+                          [||] l
+                      |> Array.to_list |> PackageManagerSpecTupleSet.of_list )
+                    |> Js.Promise.resolve))
+  end
+end
+
+type t =
+  { spec : PackageManager.spec
+  ; projectRoot : Fpath.t
+  }
+
+type resources =
+  { spec : PackageManager.spec
+  ; projectRoot : Fpath.t
+  }
+
+let init ~env ~folder =
+  let projectRoot = Fpath.ofString folder in
+  Js.Promise.all2
+    ( PackageManager.available ~env
+    , PackageManager.alreadyUsed projectRoot
+      |> Js.Promise.then_ (function
+           | Ok [] ->
+             PackageManager.Manifest.lookup projectRoot
+             |> okThen (fun x ->
+                    x |> PackageManager.PackageManagerSpecTupleSet.elements
+                    |> function
+                    | [] -> Error "TODO: global toolchain"
+                    | packageManagersInUse -> packageManagersInUse |> R.return)
+           | Ok packageManagersInUse ->
+             packageManagersInUse
+             |> List.map (fun x -> (x, projectRoot))
+             |> R.return |> Js.Promise.resolve
+           | Error msg -> Error msg |> Js.Promise.resolve) )
+  |> Js.Promise.then_
+       (fun (availablePackageManagers, alreadyUsedPackageManagers) ->
+         let availablePackageManagers =
+           match availablePackageManagers with
+           | Ok x -> x
+           | Error msg ->
+             Js.log2 "Error during availablePackageManagers()" msg;
+             []
+         in
+         let alreadyUsedPackageManagers =
+           match alreadyUsedPackageManagers with
+           | Ok x -> x
+           | Error msg ->
+             Js.log2 "Error during alreadyUsedPackageManagers()" msg;
+             []
+         in
+         match
+           List.filter
+             (fun (x, _) ->
+               match
+                 List.find_opt (fun y -> y = x) availablePackageManagers
```

Comment:
Is this just an obfuscated version of `List.exists`?

Reply:
This should have been a set intersection operation. Fixed it